File: src/capabilities.ts
import { digestCanonical, type Digest } from "./digest";
import { AlgalError } from "./errors";
import { asSafeId, type JsonValue } from "./values";

export type CapabilityHandle = `cap:${string}:${Digest}`;

const HANDLE = /^cap:([a-z][a-z0-9-]{0,63}):(sha256:[0-9a-f]{64})$/;

export function asCapabilityClass(value: unknown, what = "capability class"): string {
  return asSafeId(value, what);
}

export function capabilityHandle(
  capability: string,
  descriptor: JsonValue,
): CapabilityHandle {
  const checked = asCapabilityClass(capability);
  return `cap:${checked}:${digestCanonical(descriptor)}`;
}

export function parseCapabilityHandle(
  value: unknown,
  expected?: string,
  what = "capability handle",
): { handle: CapabilityHandle; capability: string; digest: Digest } {
  if (typeof value !== "string") {
    throw new AlgalError("TYPE_MISMATCH", `${what} must be an opaque capability handle`);
  }
  const match = HANDLE.exec(value);
  if (!match) {
    throw new AlgalError("TYPE_MISMATCH", `${what} is not a valid capability handle`);
  }
  const capability = match[1]!;
  if (expected !== undefined && capability !== expected) {
    throw new AlgalError(
      "TYPE_MISMATCH",
      `${what} carries ${capability}, expected ${expected}`,
    );
  }
  return {
    handle: value as CapabilityHandle,
    capability,
    digest: match[2] as Digest,
  };
}

File: crates/algal/src/capabilities.rs
use crate::{
    Error, Result,
    canonical::{check_digest, digest},
    contract::id,
};
use serde_json::{Value, json};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CapabilityHandle {
    pub handle: String,
    pub capability: String,
    pub digest: String,
}

pub fn capability_handle(capability: &str, descriptor: &Value) -> Result<String> {
    id(&json!(capability))?;
    Ok(format!("cap:{capability}:{}", digest(descriptor)?))
}

pub fn parse_capability_handle(handle: &str, expected: Option<&str>) -> Result<CapabilityHandle> {
    let parts: Vec<_> = handle.split(':').collect();
    if parts.len() != 4 || parts[0] != "cap" || parts[2] != "sha256" {
        return Err(Error::new("TYPE_MISMATCH", "invalid capability handle"));
    }
    id(&json!(parts[1]))?;
    let digest = format!("sha256:{}", parts[3]);
    check_digest(&digest)?;
    if expected.is_some_and(|wanted| wanted != parts[1]) {
        return Err(Error::new(
            "TYPE_MISMATCH",
            format!(
                "capability carries {}, expected {}",
                parts[1],
                expected.unwrap_or_default()
            ),
        ));
    }
    Ok(CapabilityHandle {
        handle: handle.to_owned(),
        capability: parts[1].to_owned(),
        digest,
    })
}
